dipy/reconst/ivim.py

```diff
@@ -0,0 +1,391 @@
+        qform = vec_val_vect(evecs, evals)
+
+        sphere = Sphere(xyz=gtab.bvecs[~gtab.b0s_mask])
+        ADC = apparent_diffusion_coef(qform, sphere)
+
+        S0_hat = np.mean(data[..., ~gtab.b0s_mask] /
+                         np.exp(-gtab.bvals[~gtab.b0s_mask] * ADC),
+                         -1)
+        return np.array([S0_hat, D_guess])
+
+    def _leastsq(self, data, x0):
+        """
+        Use leastsq for finding ivim_params
+
+        Parameters
+        ----------
+        data : array, (len(bvals))
+            An array containing the signal from a voxel.
+            If the data was a 3D image of 10x10x10 grid with 21 bvalues,
+            the multi_voxel decorator will run the single voxel fitting
+            on all the 1000 voxels to get the parameters in
+            IvimFit.model_paramters. The shape of the parameter array
+            will be (data[:-1], 4).
+
+        x0 : array
+            Initial guesses for the parameters S0, f, D_star and D
+            calculated using the function `_estimate_S0_D`
+        """
+        gtol = self.options["gtol"]
+        ftol = self.options["ftol"]
+        xtol = self.tol
+        epsfcn = self.options["eps"]
+        maxfev = self.options["maxiter"]
+        bounds = self.bounds
+        bvals = self.gtab.bvals
+        if not SCIPY_LESS_0_17:
+            res = least_squares(_ivim_error,
+                                x0,
+                                bounds=bounds,
+                                ftol=ftol,
+                                xtol=xtol,
+                                gtol=gtol,
+                                max_nfev=maxfev,
+                                args=(bvals, data))
+            ivim_params = res.x
+            return ivim_params
+        else:
+            res = leastsq(_ivim_error,
+                          x0,
+                          args=(bvals, data),
+                          gtol=gtol,
+                          xtol=xtol,
+                          ftol=ftol,
+                          epsfcn=epsfcn,
+                          maxfev=maxfev)
+            ivim_params = res[0]
+            return ivim_params
+
+
+class IvimFit(object):
+
+    def __init__(self, model, model_params):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            model : Model class
+            model_params : array
+                The parameters of the model. In this case it is an
+                array of ivim parameters. If the fitting is done
+                for multi_voxel data, the multi_voxel decorator will
+                run the fitting on all the voxels and model_params
+                will be an array of the dimensions (data[:-1], 4),
+                i.e., there will be 4 parameters for each of the voxels.
+        """
+        self.model = model
+        self.model_params = model_params
+
+    def __getitem__(self, index):
+        model_params = self.model_params
```

Comment:
This method is not currently being used in the tests. Please write a test that uses it. Be sure to test the corner cases (e.g., that an `IndexError` gets raised for invalid indices). Be sure to test this both in the single and multi voxel case.